### PR TITLE
add type annotations #18

### PR DESCRIPTION
**Pull request by matt-land**
Adding type hints for functions. This does not currently pass all mypy. The  4 remaining issues are as follows (and beyond my expertise):
```
mypy .
adafruit_bitmapsaver.py:101: error: Unsupported operand types for & ("None" and "int")
adafruit_bitmapsaver.py:101: note: Left operand is of type "Union[int, None, Any]"
adafruit_bitmapsaver.py:102: error: Unsupported operand types for >> ("None" and "int")
adafruit_bitmapsaver.py:102: note: Left operand is of type "Union[int, None, Any]"
adafruit_bitmapsaver.py:109: error: Value of type "Union[Any, bytearray, memoryview, array[Any], ndarray, RGBMatrix]" is not indexable
adafruit_bitmapsaver.py:138: error: Module has no attribute "DISPLAY"
Found 4 errors in 1 file (checked 1 source file)

### CODE REVIEW

File: adafruit_bitmapsaver.py
Comment (FoamyGuy) [inline adafruit_bitmapsaver.py:38]: I think maybe instead of `BufferedWriter` we may want `WriteableBuffer` from circuitpython_typing but I'm not 100% certain. @tekktrik do you think `WriteableBuffer` makes sense as used in this library?

File: adafruit_bitmapsaver.py
Comment (FoamyGuy) [inline adafruit_bitmapsaver.py:100]: lets remove the `#type: ignore`'s for now. We haven't added mypy to the CI yet so it's okay if it's still raising errors when run locally. Right now we're primarily focused on the "missing type annotations" error for function arguments and returns. The others we'll get around to when we do add mypy to CI and at that time we'll figure out what configuration we want to use which will guide which things if any need the ignore comments.

File: adafruit_bitmapsaver.py
Comment (FoamyGuy) [inline adafruit_bitmapsaver.py:88]: Is palette really optional here? it doesn't default to `None` and it looks like there is an attempt to access a value from it using square brackets inside of this function without any further check for None that I'm seeing.

File: adafruit_bitmapsaver.py
Comment (FoamyGuy) [inline adafruit_bitmapsaver.py:107]: same here, I think we can remove the mypy ignore comments for now

File: mypy.ini
Comment (FoamyGuy) [inline mypy.ini:8]: I don't think we are ready to add the mypy configuration file yet. Checking in with @kattni too though, do you have preference on adding this file now vs. waiting until whenever we are ready to enable mypy in our CI?

File: adafruit_bitmapsaver.py
Comment (FoamyGuy) [inline adafruit_bitmapsaver.py:136]: it was written with `dir()` originally but it's probably better if this used `hasattr` instead. Do you mind making that change while you're working on this?
so instead of:
```
if "DISPLAY" not in dir(board):
```
it could be:
```
if not hasattr(board, 'DISPLAY'):
```

**Review by FoamyGuy — COMMENTED** [review]
Thanks for working on this @matt-land I noted a few potential changes in some different sections of the code, but this is looking pretty good to me overall.

File: adafruit_bitmapsaver.py
Comment (matt-land) [reply adafruit_bitmapsaver.py:88]: Palette is optional. The function takes either a (bitmap & palette) and follows a major branch or (Display & None) and follows a different major branch

Related: I added and "# handled by save_pixel's guardians" after I figured this out

File: mypy.ini
Comment (matt-land) [reply mypy.ini:8]: I can pull it, NBD. But as a (sometimes) maintainer, I'd be pleased it if it went in, because if/ when the next person works here (with types), they would immediately need to recreate this file (or another file providing similar functionality).

I don't think this mypy.ini is ready to go to Cookie Cutter for all repos yet:
- run 5-6 projects and see how similar their mypy.ini files. Bake a  common target config file from learnings.
- standardize the target config in mypy.ini, pre-commit hooks, or setup.cfg

File: adafruit_bitmapsaver.py
Comment (FoamyGuy) [inline adafruit_bitmapsaver.py:125]: This `Palette` should be optional I think since it's default value is None. 

I can add a commit with that change. It's looking ready to go after that to me.

**Review by FoamyGuy — APPROVED** [review]
Looks good to me. Thanks again @matt-land 

We'll keep the mypy ini file for now and adjust as needed when something is rolled out across all of the libraries.

**Pull request merged**
